lib/collection/src/shards/shard_holder.rs

```diff
@@ -561,11 +592,40 @@ impl ShardHolder {
             }
         }
 
+        // NOTE:
+        //
+        // `keep` calls are *sync*, so once `move_file` future resolves, `keep` calls will *always*
+        // execute to completion and won't be "cancelled" even if `create_shard_snapshot` future is
+        // dropped.
+        //
+        // If we switch to *async* `keep` calls at some point:
+        // - we may leave it as-is, but files may get cleaned up or persisted semi-randomly in that case
+        //   - (but do we care, if request was cancelled?)
+        // - or we can spawn `move_file` and `keep` calls as a single task
+        //   - so that `keep` calls will *always* execute to completion if `move_file` resolves
+        //   - and we can do "select" on `move_file` and a cancellation token, to be able to cancel
+        //     `move_file` before it resolves
+
+        // `tempfile::NamedTempFile::persist` does not work if destination file is on another
+        // file-system, so we have to move the file explicitly.
         move_file(temp_file.path(), &snapshot_path).await?;
 
+        // We already moved the file to the desired destination, but `temp_file` will still try to
+        // delete the file when dropped, so we use `tempfile::NamedTempFile::keep` to consume it safely.
+        //
+        // `tempfile::NamedTempFile::keep` "needs to mark the file as non-temporary" on Windows,
+        // which is a fallible operation. But we already moved the file, so we expect it to fail
+        // and explicitly ignore the result.
+        //
+        // We may mark the wrong file as "non-temporary", if someone creates another file with the
+        // same name in between `move_file` and `tempfile::NamedTempFile::keep` calls, but this is
+        // better than *deleting* it, if we drop `temp_file` as-is
```

Comment:
I don't think we need all this extra text

```suggestion
        // Temp file is moved away and is persisted, don't delete it anymore
```